In [67]:
import graphlab

In [68]:
image_train=graphlab.SFrame('C:\Users\ROHIT\Desktop\MACHINE LEARNING\COURSE 1\WEEK 6\DATA\Week 6\image_train_data')
image_test=graphlab.SFrame('C:\Users\ROHIT\Desktop\MACHINE LEARNING\COURSE 1\WEEK 6\DATA\Week 6\image_test_data')

# Q1. What’s the least common category in the training data? 

In [69]:
sketch = graphlab.Sketch(image_train['label'])
sketch = image_train['label'].sketch_summary()

In [70]:
sketch


+------------------+-------+----------+
|       item       | value | is exact |
+------------------+-------+----------+
|      Length      |  2005 |   Yes    |
| # Missing Values |   0   |   Yes    |
| # unique values  |   4   |    No    |
+------------------+-------+----------+

Most frequent items:
+-------+------------+-----+-----+------+
| value | automobile | cat | dog | bird |
+-------+------------+-----+-----+------+
| count |    509     | 509 | 509 | 478  |
+-------+------------+-----+-----+------+


# 2. Creating category-specific image retrieval models: In most retrieval tasks, the data we have is unlabeled, thus we call these unsupervised learning problems. However, we have labels in this image dataset, and will use these to create one model for each of the 4 image categories, {‘dog’,’cat’,’automobile’,bird’}. To start, follow these steps:

Split the SFrame with the training data into 4 different SFrames. Each of these will contain data for 1 of the 4 categories above. Hint: if you use a logical filter to select the rows where the ‘label’ column equals ‘dog’, you can create an SFrame with only the data for images labeled ‘dog’.Similarly to the image retrieval notebook you downloaded, you are going to create a nearest neighbor model using the 'deep_features' as the features, but this time create one such model for each category, using the corresponding subset of the training_data. You can call the model with the ‘dog’ data the dog_model, the one with the ‘cat’ data the cat_model, as so on.

You now have a nearest neighbors model that can find the nearest ‘dog’ to any image you give it, the dog_model; one that can find the nearest ‘cat’, the cat_model; and so on.

Using these models, answer the following questions. The cat image below is the first in the test data:

In [71]:
dog=image_train.filter_by(['dog'],'label')

In [72]:
cat=image_train.filter_by(['cat'],'label')

In [73]:
automobile=image_train.filter_by(['automobile'],'label')

In [74]:
bird=image_train.filter_by(['bird'],'label')

## Creating a nearest neighbor model for each category

In [75]:
dog_model=graphlab.nearest_neighbors.create(dog,
                                           features=['deep_features'],
                                            label='id')

Starting brute force nearest neighbors model training.

In [76]:
cat_model=graphlab.nearest_neighbors.create(cat,
                                           features=['deep_features'],
                                            label='id')

Starting brute force nearest neighbors model training.

In [77]:
automobile_model=graphlab.nearest_neighbors.create(automobile,
                                           features=['deep_features'],
                                            label='id')

Starting brute force nearest neighbors model training.

In [78]:
bird_model=graphlab.nearest_neighbors.create(bird,
                                           features=['deep_features'],
                                            label='id')

Starting brute force nearest neighbors model training.

In [97]:
graphlab.canvas.set_target('browser')
image_test[0:1]

Columns:
	id	int
	image	Image
	label	str
	deep_features	array
	image_array	array

Rows: 1

Data:
+----+----------------------+-------+-------------------------------+
| id |        image         | label |         deep_features         |
+----+----------------------+-------+-------------------------------+
| 0  | Height: 32 Width: 32 |  cat  | [1.13469004631, 0.0, 0.0, ... |
+----+----------------------+-------+-------------------------------+
+-------------------------------+
|          image_array          |
+-------------------------------+
| [158.0, 112.0, 49.0, 159.0... |
+-------------------------------+
[1 rows x 5 columns]

In [80]:
image_test[0:1]['image'].show()

Canvas is accessible via web browser at the URL: http://localhost:49919/index.html
Opening Canvas in default web browser.


In [98]:
cat=image_test[0:1]

In [99]:
cat_model.query(cat)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 13.034ms     |

| Done         |         | 100         | 116.813ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	int
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      16289      |  34.623719208 |  1   |
|      0      |      45646      | 36.0068799284 |  2   |
|      0      |      32139      | 36.5200813436 |  3   |
|      0      |      25713      | 36.7548502521 |  4   |
|      0      |       331       | 36.8731228168 |  5   |
+-------------+-----------------+---------------+------+
[5 rows x 4 columns]

In [100]:
def get_images_from_ids(query_result):
    return image_train.filter_by(query_result['reference_label'],'id')

In [101]:
cat_neighbors=get_images_from_ids(cat_model.query(cat))

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 11.027ms     |

| Done         |         | 100         | 86.349ms     |

+--------------+---------+-------------+--------------+

In [102]:
cat_neighbors['image'].show()

Canvas is accessible via web browser at the URL: http://localhost:49919/index.html
Opening Canvas in default web browser.


### 
What is the nearest ‘dog’ labeled image in the training data to the cat image above (the first image in the test data)?

In [87]:
dog=image_test[0:1]

In [88]:
dog_model.query(dog)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 13.536ms     |

| Done         |         | 100         | 97.263ms     |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	int
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      16976      | 37.4642628784 |  1   |
|      0      |      13387      | 37.5666832169 |  2   |
|      0      |      35867      | 37.6047267079 |  3   |
|      0      |      44603      | 37.7065585153 |  4   |
|      0      |       6094      | 38.5113254907 |  5   |
+-------------+-----------------+---------------+------+
[5 rows x 4 columns]

In [93]:
def get_images_from_ids(query_result):
    return image_train.filter_by(query_result['reference_label'],'id')

In [96]:
get_images_from_ids(dog_model.query(dog))['image'].show()

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 11.03ms      |

| Done         |         | 100         | 109.45ms     |

+--------------+---------+-------------+--------------+

Canvas is accessible via web browser at the URL: http://localhost:49919/index.html
Opening Canvas in default web browser.


  # Q3. A simple example of nearest-neighbors classification: When we queried a nearest neighbors model, the ‘distance’ column in the table above shows the computed distance between the input and each of the retrieved neighbors. In this question, you will use these distances to perform a classification task, using the idea of a nearest-neighbors classifier.

For the first image in the test data (image_test[0:1]), which we used above, compute the mean distance between this image at its 5 nearest neighbors that were labeled ‘cat’ in the training data (similarly to what you did in the previous question). Save this result.Similarly, for the first image in the test data (image_test[0:1]), which we used above, compute the mean distance between this image at its 5 nearest neighbors that were labeled ‘dog’ in the training data (similarly to what you did in the previous question). Save this result.On average, is the first image in the test data closer to its 5 nearest neighbors in the ‘cat’ data or in the ‘dog’ data? (In a later course, we will see that this is an example of what is called a k-nearest neighbors classifier, where we use the label of neighboring points to predict the label of a test point.)

In [103]:
cat_model.query(image_test[0:1])['distance'].mean()

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 21.557ms     |

| Done         |         | 100         | 94.751ms     |

+--------------+---------+-------------+--------------+

36.15573070978294

In [104]:
dog_model.query(image_test[0:1])['distance'].mean()

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 13.536ms     |

| Done         |         | 100         | 111.296ms    |

+--------------+---------+-------------+--------------+

37.77071136184156

# Q4. [Challenging Question] Computing nearest neighbors accuracy using SFrame operations:  
### A nearest neighbor classifier predicts the label of a point as the most common label of its nearest neighbors. In this question, we will measure the accuracy of a 1-nearest-neighbor classifier, i.e., predict the output as the label of the nearest neighbor in the training data. Although there are simpler ways of computing this result, we will go step-by-step here to introduce you to more concepts in nearest neighbors and SFrames, which will be useful later in this Specialization.

## Step 1: Training models: For this question, you will need the nearest neighbors models you learned above on the training data, i.e., the dog_model, cat_model, automobile_model and bird_model.Spliting test data by label: Above, you split the train data SFrame into one SFrame for images labeled ‘dog’, another for those labeled ‘cat’, etc. Now, do the same for the test data. You can call the resulting SFrames

In [105]:
image_test_dog=image_test.filter_by(['dog'],'label')

In [106]:
image_test_cat=image_test.filter_by(['cat'],'label')

In [107]:
image_test_automobile=image_test.filter_by(['automobile'],'label')

In [108]:
image_test_bird=image_test.filter_by(['bird'],'label')

## Step2: Finding nearest neighbors in the training set for each part of the test set: Thus far, we have queried, e.g.,



our nearest neighbors models with a single image as the input, but you can actually query with a whole set of data, and it will find the nearest neighbors for each data point. Note that the input index will be stored in the ‘query_label’ column of the output SFrame.

Using this knowledge find the closest neighbor in to the dog test data using each of the trained models, e.g.,



In [113]:
dog_cat_neighbors = cat_model.query(image_test_dog, k=1)

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 616.717ms    |

| Done         | 509000  | 100         | 652.355ms    |

+--------------+---------+-------------+--------------+

In [114]:
#[‘dog-dog’] 
dog_dog_neighbors = dog_model.query(image_test_dog, k=1)

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 572.523ms    |

| Done         | 509000  | 100         | 599.593ms    |

+--------------+---------+-------------+--------------+

In [115]:
#[‘dog-automobile’] 

dog_automobile_neighbors = automobile_model.query(image_test_dog, k=1)

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 567.01ms     |

| Done         | 509000  | 100         | 625.662ms    |

+--------------+---------+-------------+--------------+

In [116]:
#[‘dog-bird’] 

dog_bird_neighbors = bird_model.query(image_test_dog, k=1)

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 120000  | 25.1046     | 541.943ms    |

| Done         | 478000  | 100         | 628.193ms    |

+--------------+---------+-------------+--------------+

# Create an SFrame with the distances from ‘dog’ test examples to the respective nearest neighbors in each class in the training data: The ‘distance’ column in dog_cat_neighbors above contains the distance between each ‘dog’ image in the test set and its nearest ‘cat’ image in the training set. The question we want to answer is how many of the test set ‘dog’ images are closer to a ‘dog’ in the training set than to a ‘cat’, ‘automobile’ or ‘bird’. So, next we will create an SFrame containing just these distances per data point. The goal is to create an SFrame called dog_distances with 4 columns:

i. dog_distances[‘dog-dog’] ---- storing dog_dog_neighbors[‘distance’]

ii. dog_distances[‘dog-cat’] ---- storing dog_cat_neighbors[‘distance’]

iii. dog_distances[‘dog-automobile’] ---- storing dog_automobile_neighbors[‘distance’]

iv. dog_distances[‘dog-bird’] ---- storing dog_bird_neighbors[‘distance’]

Hint: You can create a new SFrame from the columns of other SFrames by creating a dictionary with the new columns, as shown in this example:

In [117]:
dog_cat_neighbors

Columns:
	query_label	int
	reference_label	int
	distance	float
	rank	int

Rows: 1000

Data:
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |        33       | 36.4196077068 |  1   |
|      1      |      30606      | 38.8353268874 |  1   |
|      2      |       5545      | 36.9763410854 |  1   |
|      3      |      19631      | 34.5750072914 |  1   |
|      4      |       7493      |  34.778824791 |  1   |
|      5      |      47044      | 35.1171578292 |  1   |
|      6      |      13918      | 40.6095830913 |  1   |
|      7      |      10981      | 39.9036867306 |  1   |
|      8      |      45456      | 38.0674700168 |  1   |
|      9      |      44673      | 42.7258732951 |  1   |
+-------------+-----------------+---------------+------+
[1000 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

### new_sframe = graphlab.SFrame({'foo': other_sframe['foo'],'bar': some_other_sframe['bar']})

In [118]:
dog_distances = graphlab.SFrame({'dog_automobile': dog_automobile_neighbors['distance'],
                              'dog_bird': dog_bird_neighbors['distance'],
                              'dog_cat': dog_cat_neighbors['distance'],
                              'dog_dog': dog_dog_neighbors['distance']
                             })

In [119]:
dog_distances.head()

Columns:
	dog_automobile	float
	dog_bird	float
	dog_cat	float
	dog_dog	float

Rows: 10

Data:
+----------------+---------------+---------------+---------------+
| dog_automobile |    dog_bird   |    dog_cat    |    dog_dog    |
+----------------+---------------+---------------+---------------+
| 41.9579761457  | 41.7538647304 | 36.4196077068 | 33.4773590373 |
| 46.0021331807  | 41.3382958925 | 38.8353268874 | 32.8458495684 |
| 42.9462290692  | 38.6157590853 | 36.9763410854 | 35.0397073189 |
| 41.6866060048  | 37.0892269954 | 34.5750072914 | 33.9010327697 |
| 39.2269664935  |  38.272288694 |  34.778824791 | 37.4849250909 |
| 40.5845117698  | 39.1462089236 | 35.1171578292 |  34.945165344 |
| 45.1067352961  |  40.523040106 | 40.6095830913 | 39.0957278345 |
| 41.3221140974  | 38.1947918393 | 39.9036867306 | 37.7696131032 |
| 41.8244654995  | 40.1567131661 | 38.0674700168 | 35.1089144603 |
| 45.4976929401  | 45.5597962603 | 42.7258732951 | 43.2422832585 |
+----------------+---------------+---------------+---------------+
[10 rows x 4 columns]

### Computing the number of correct predictions using 1-nearest neighbors for the dog class: Now that you have created the SFrame dog_distances, you will learn to use the method
.apply()
on this SFrame to iterate line by line and compute the number of ‘dog’ test examples where the distance to the nearest ‘dog’ was lower than that to the other classes. You will do this in three steps:

i. Consider one row of the SFrame dog_distances. Let’s call this variable row. You can access each distance by calling, for example,

row[‘dog_cat’]
which, in example table above, will have value equal to 36.4196077068 for the first row.

Create a function starting with

def is_dog_correct(row):
which returns 1 if the value for row[‘dog_dog’] is lower than that of the other columns, and 0 otherwise. That is, returns 1 if this row is correctly classified by 1-nearest neighbors, and 0 otherwise

In [120]:
def is_dog_correct(row):  
    if row['dog_dog'] <= min(row.values()):     
        return 1    
    else:        
        return 0

# dog_distances.apply(lambda row: 1 if row['dog_dog'] <= min(row.values()) else 0)

In [123]:
dog_distances.apply(is_dog_correct)

Less than 4 successfully started. Using only 3 workers.

All operations will proceed as normal, but lambda operations will not be able to use all available cores.

(The location of the log file is printed at the start of the GraphLab server).

dtype: int
Rows: 1000
[1L, 1L, 1L, 1L, 0L, 1L, 1L, 1L, 1L, 0L, 1L, 0L, 0L, 1L, 1L, 1L, 1L, 0L, 0L, 1L, 0L, 0L, 0L, 0L, 1L, 1L, 0L, 1L, 1L, 0L, 1L, 1L, 0L, 1L, 1L, 1L, 1L, 1L, 0L, 0L, 1L, 1L, 0L, 1L, 1L, 0L, 1L, 1L, 1L, 1L, 0L, 0L, 1L, 1L, 0L, 1L, 1L, 1L, 0L, 0L, 1L, 0L, 0L, 1L, 0L, 1L, 0L, 0L, 0L, 1L, 1L, 1L, 1L, 0L, 0L, 0L, 1L, 1L, 1L, 1L, 1L, 1L, 0L, 0L, 1L, 1L, 1L, 1L, 1L, 1L, 1L, 0L, 1L, 1L, 1L, 1L, 1L, 1L, 0L, 0L, ... ]

In [124]:
dog_distances.apply(is_dog_correct).sum()

678L

### 
Hint: To make sure your code is working correctly, if you were to do steps d) and e) in this question to count the number of correctly classified ‘cat’ images in the test data, instead of ‘dog’, the result would be 548.

In [125]:
cat_distances = graphlab.SFrame({'cat_automobile': automobile_model.query(image_test_cat, k=1)['distance'],
                                 'cat_bird': bird_model.query(image_test_cat, k=1)['distance'],
                                 'cat_cat': cat_model.query(image_test_cat, k=1)['distance'],
                                 'cat_dog': dog_model.query(image_test_cat, k=1)['distance'],
                                })

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 555.98ms     |

| Done         | 509000  | 100         | 612.629ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 120000  | 25.1046     | 667.841ms    |

| Done         | 478000  | 100         | 702.432ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 567.51ms     |

| Done         | 509000  | 100         | 613.171ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 732.447ms    |

| Done         | 509000  | 100         | 907.508ms    |

+--------------+---------+-------------+--------------+

In [126]:
cat_distances.head()

Columns:
	cat_automobile	float
	cat_bird	float
	cat_cat	float
	cat_dog	float

Rows: 10

Data:
+----------------+---------------+---------------+---------------+
| cat_automobile |    cat_bird   |    cat_cat    |    cat_dog    |
+----------------+---------------+---------------+---------------+
| 39.6710582792  |  38.074265869 |  34.623719208 | 37.4642628784 |
| 43.0089056688  | 36.3674024138 | 33.8680579302 | 29.3472319585 |
| 38.6010006604  | 35.3039394947 | 32.4615168902 | 32.2599640475 |
| 39.3566307091  | 38.8944029601 | 35.7708210254 | 35.3852085188 |
| 38.3572372618  | 34.2820409875 | 31.1577686417 | 30.0442985088 |
| 42.0904793181  | 44.5352170178 | 41.3986035847 | 35.4741000424 |
| 39.0520251253  | 34.0290595084 | 30.9894594959 | 32.5845275226 |
| 39.3058645069  | 39.0236924983 | 37.0814607387 | 37.6502852614 |
| 43.0248129799  | 40.8334054297 | 39.9883863688 | 36.9801353512 |
| 45.6749176426  | 40.1258835601 | 39.7076633097 | 41.1259410707 |
+----------------+---------------+---------------+---------------+
[10 rows x 4 columns]

In [131]:
def is_cat_correct(row):  
    if row['cat_cat'] <= min(row.values()):     
        return 1    
    else:        
        return 0
    

In [132]:
cat_distances.apply(is_cat_correct).sum()

548L

# 
Accuracy of predicting dog in the test data: Using the work you did in this question, what is the accuracy of the 1-nearest neighbor classifier at classifying ‘dog’ images from the test set?

In [133]:
dog_distances.apply(is_dog_correct).sum()/float(len(dog_distances))

0.678